File: 493-reverse-pairs/reverse-pairs.cpp
class Solution {
public:

    void merge(vector<int>&nums,int st,int mid,int en){

        int i = st,j=mid;
        int v[en-st];
        int k = 0;
        for(;i<mid&&j<en;){
            if(nums[i]<=nums[j]){
                v[k] = nums[i];k++;
                i++;
            }else{
                v[k] = nums[j];k++;
                j++;
            }

        }
        while(i<mid){
           v[k] = nums[i];k++;
            i++;
        }

        while(j<en){
            v[k] = nums[j];k++;
            j++;
        }
        i = st;
        for( k=0;k<en-st;k++){
            nums[i] = v[k];i++;
        }


    }

    int  mergesort(vector<int>&nums ,int st,int en){

        if(st+1==en || st==en){return 0;}
        int mid = (st+en)/2;
        int ans = mergesort(nums,st,mid) + mergesort(nums,mid,en);

        for(int i=st,j=mid;i<mid;i++){
            while(j<en&&nums[i]/2.0>nums[j]){
                j++;
            }
            ans += j-mid;
        }

        merge(nums,st,mid,en);
        return ans;
    }

    int reversePairs(vector<int>& nums) {
        int ans;
        ans = mergesort(nums,0,nums.size());
        return ans;
    }
};

File: 493-reverse-pairs/reverse-pairs.java
class Solution {
    public void merge(int[] nums, int st,int en,int mid){
        ArrayList<Integer> li = new ArrayList<>();
        
        int i = st,j=mid;
        while(i<mid && j<en){
            if(nums[i]<nums[j]){
                li.add(nums[i]);i++;
            }else{
                li.add(nums[j]);j++;
            }
        }
        while(i<mid){
            li.add(nums[i]);i++;
        }
        while(j<en){
            li.add(nums[j]);j++;
        }

        i = st;
        for(int num: li){
            nums[i]=num;i++;
        }
        return;
    }

    public void mergesort(int[] nums,int st,int en,int[] ans){
        if(st==en-1){
            return;
        }
        int mid = (st+en)/2;
        mergesort(nums,st,mid,ans);
        mergesort(nums,mid,en,ans);

        for(int i=st,j=mid;i<mid&&j<en;){
            if(nums[i]/2.0 > nums[j]){
                ans[0] += (mid-i);
                j++;
            }else{
                i++;
            }
        }

        merge(nums,st,en,mid);
    }

    public int reversePairs(int[] nums) {
        int[] ans = new int[1];
        mergesort(nums,0,nums.length,ans);
        return ans[0];
    }
}